### ISSUE TITLE
Check handling of multi-line attribute instantiations by the function call sniffs

### ISSUE THREAD

**Issue by jrfnl** [PHPCSStandards/PHP_CodeSniffer#1281]
## Setting the scene

PHP 8.0 introduced support for attributes via the `#[...]` syntax. At this moment, neither PHPCS itself, nor PHPCSExtra contain any sniffs to handle the formatting of attributes.

The [PER Coding Standard](https://www.php-fig.org/per/coding-style/) from FIG, since PER 2.0, outlines a set of rules for attribute formatting to comply with, so using those rules as a starting point would allow for creating an initial set of sniffs to address attribute formatting.

## The PER rule

[From PER](https://www.php-fig.org/per/coding-style/#123-compound-attributes):
> The attribute arguments MUST follow the same rules as defined for multiline function calls.

The rules for (multiline) function calls are covered in https://www.php-fig.org/per/coding-style/#47-method-and-function-calls

## The action

The above rules should probably be covered via pre-existing function call sniffs.

It will need investigating which sniff(s) should handle this and whether these sniffs already handle attribute instantiations or not.

This includes checking things like whether the indentation of multi-line attribute instantiations is checked by the function call sniffs or the `ScopeIndent` sniff, whether parameters should be on one line or multiple lines is handled by the sniff etc.

The initial outcome should be a list of sniff names with what part of the PER multi-line function call rules are handled by each and whether those rules are also (correctly) applied for attribute class instantiations.

The next step after that would be updating the sniffs where needed.

Please note: the initial investigation and the sniff updates do not have to be done by the same contributor. These are separate tasks.

**Comment by Amoifr** [comment]
Here is the initial investigation this asks for. Tested against PHP_CodeSniffer 4.0.4 on PHP 8.5, with fixtures that pair each PER §4.7 rule with the equivalent violation written as a plain function call, so that "does the sniff see the attribute" is separated from "does the sniff have the rule at all".

## What already works

`PEAR.Functions.FunctionCallSignature` and `Generic.Functions.FunctionCallArgumentSpacing` treat an attribute instantiation exactly like a function call. Every case below reports on the attribute and on its function-call control, with the same error code:

| PER §4.7 rule | Sniff and error code | Function call | Attribute |
| --- | --- | :---: | :---: |
| No space before `(` | `PEAR…FunctionCallSignature.SpaceBeforeOpenBracket` | yes | yes |
| No space after `(` | `PEAR…FunctionCallSignature.SpaceAfterOpenBracket` | yes | yes |
| No space before `)` | `PEAR…FunctionCallSignature.SpaceBeforeCloseBracket` | yes | yes |
| Split args: `(` last on its line | `PEAR…FunctionCallSignature.ContentAfterOpenBracket` | yes | yes |
| Split args: one indent level | `PEAR…FunctionCallSignature.Indent` | yes | yes |
| Split args: `)` on its own line | `PEAR…FunctionCallSignature.CloseBracketLine` | yes | yes |
| Space after each comma | `Generic…FunctionCallArgumentSpacing.NoSpaceAfterComma` | yes | yes |

`PSR2.Methods.FunctionCallSignature` extends the PEAR sniff, so `PSR12` reports the same set.

Two answers to the specific questions in the issue:

- **Indentation is handled by the function call sniff, not `ScopeIndent`.** Run in isolation on the failing fixture, `Generic.WhiteSpace.ScopeIndent` reports nothing at all; every indentation error comes from `PEAR…FunctionCallSignature.Indent`.
- **A PER-conforming multi-line attribute passes cleanly**, both at file scope and on a class member. No false positives to fix.

One thing I chased and discarded, so nobody else spends time on it: when the closing parenthesis is left on the last argument's line, the `Indent` message reads `expected 0 spaces but found 4` rather than `expected 4`. That looked attribute-specific, but the identical function call at the same scope produces the identical message, so it is a quirk of how the sniff reports that line and not an attribute problem.

## What nothing covers

The rules with no function-call equivalent are unhandled, by these sniffs and by `PSR12` as a whole. None of the following is reported:

```php
#[Attr1('a') ,Attr2('b')]   // §12.3 compound attributes, separator spacing
#[Attr3] class SameLine {}  // §12 the attribute should be on its own line
#[ Attr4('a') ]             // spacing just inside #[ ]
```

So the split for the follow-up work looks like: the multi-line argument formatting of PER §4.7 needs no sniff changes, and a new sniff would be needed for the attribute-specific part of §12.

Happy to take that second part on if you want it, and happy to be corrected on any of the above.
